pkg/ruler/rule_concurrency.go

```diff
@@ -0,0 +1,212 @@
+			Name: "cortex_ruler_independent_rule_evaluation_concurrency_attempts_started_total",
+			Help: "Total number of started attempts to acquire concurrency slots across all tenants",
+		}),
+		AcquireFailedTotal: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: "cortex_ruler_independent_rule_evaluation_concurrency_attempts_incomplete_total",
+			Help: "Total number of incomplete attempts to acquire concurrency slots across all tenants",
+		}),
+		ReleaseTotal: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: "cortex_ruler_independent_rule_evaluation_concurrency_attempts_completed_total",
+			Help: "Total number of concurrency slots we're done using across all tenants",
+		}),
+	}
+}
+
+// MultiTenantConcurrencyController is a concurrency controller that limits the number of concurrent rule evaluations both global and per tenant.
+type MultiTenantConcurrencyController struct {
+	logger                   log.Logger
+	limits                   RulesLimits
+	maxGlobalConcurrency     int64
+	thresholdRuleConcurrency float64
+	metrics                  *MultiTenantConcurrencyControllerMetrics
+
+	globalConcurrency    semaphore.Weighted
+	tenantConcurrencyMtx sync.Mutex
+	tenantConcurrency    map[string]*DynamicSemaphore
+}
+
+// NewMultiTenantConcurrencyController creates a new MultiTenantConcurrencyController.
+func NewMultiTenantConcurrencyController(logger log.Logger, maxGlobalConcurrency int64, ThresholdRuleConcurrency float64, reg prometheus.Registerer, limits RulesLimits) *MultiTenantConcurrencyController {
+	return &MultiTenantConcurrencyController{
+		metrics:                  newMultiTenantConcurrencyControllerMetrics(reg),
+		logger:                   log.With(logger, "component", "concurrency-controller"),
+		maxGlobalConcurrency:     maxGlobalConcurrency,
+		thresholdRuleConcurrency: ThresholdRuleConcurrency,
+		limits:                   limits,
+		globalConcurrency:        *semaphore.NewWeighted(maxGlobalConcurrency),
+		tenantConcurrency:        make(map[string]*DynamicSemaphore),
+	}
+}
+
+// Done releases a slot from the concurrency controller.
+func (c *MultiTenantConcurrencyController) Done(ctx context.Context) {
+	c.globalConcurrency.Release(1)
+	c.metrics.CurrentConcurrency.Dec()
+	c.metrics.ReleaseTotal.Inc()
+
+	userID, err := tenant.TenantID(ctx)
+	if err != nil {
+		level.Error(c.logger).Log("msg", "error extracting org id from context while releasing a slot", "err", err)
+		return
+	}
+
+	c.tenantConcurrencyMtx.Lock()
+	tc, ok := c.tenantConcurrency[userID]
+	c.tenantConcurrencyMtx.Unlock()
+
+	if ok {
+		tc.Release()
+	} else {
+		level.Error(c.logger).Log("msg", "tenant concurrency controller: tenant not found", "tenant", userID)
+	}
+}
+
+// Allow tries to acquire a slot from the concurrency controller.
+func (c *MultiTenantConcurrencyController) Allow(ctx context.Context, group *rules.Group, rule rules.Rule) bool {
+	c.metrics.AcquireTotal.Inc()
```

Comment:
I'm still not convinced it's correct to increase it here. We're not even trying to acquire it if the group is not at risk or the rule is not independent. Shouldn't we increase this right before `if !c.globalConcurrency.TryAcquire(1) {`?

Reply:
My main concern is that when we'll try to compute the % of `cortex_ruler_independent_rule_evaluation_concurrency_attempts_incomplete_total`, we'll get a wrong number if we track the acquire total metric here, because a bunch (the majority) of metrics will not even try to acquire it due to the group not at right or the rule not being dependent.

Due to how metrics are tracked right now, we don't have a way to measure the % of acquire attempts that failed because we where at capacity (which is the most important information to me...).

Reply:
Damn, you're right - I see your point now.

Reply:
Done in https://github.com/grafana/mimir/pull/8146/commits/b08738da87e1e9bd7c4ce4e22191e17f43b2d6b6